packages/squiggle-lang/src/types/helpers.ts

```diff
@@ -0,0 +1,324 @@
+import { TDist } from "./TDist.js";
+import { TIntrinsic, tNumber, tString } from "./TIntrinsic.js";
+import { TNumberRange } from "./TNumberRange.js";
+import { TTuple } from "./TTuple.js";
+import { InferredOutputType, TTypedLambda } from "./TTypedLambda.js";
+import { tUnion, TUnion } from "./TUnion.js";
+import { tAny, TAny, Type } from "./Type.js";
+
+// This is not a method on `TTypedLambda` because the signatures could be
+// gathered from a union of typed lambdas; see `NodeCall` for details.
+export function inferOutputTypeByMultipleSignatures(
+  signatures: TTypedLambda[],
+  argTypes: Type[]
+): InferredOutputType {
+  // We're gathering all possible output types, not just the first one.
+  // Polymorphic functions in stdlib are greedy (they take the first matching
+  // signature), but during the analysis stage we don't know enough about the
+  // value to decide which signature will be used in runtime.
+  const possibleOutputTypes: Type[] = [];
+
+  const arityErrors: Extract<InferredOutputType, { kind: "arity" }>[] = [];
+  const noMatchErrors: Extract<InferredOutputType, { kind: "no-match" }>[] = [];
+
+  for (const signature of signatures) {
+    const outputType = signature.inferOutputType(argTypes);
+    switch (outputType.kind) {
+      case "ok":
+        possibleOutputTypes.push(outputType.type);
+        break;
+      case "arity":
+        arityErrors.push(outputType);
+        break;
+      case "no-match":
+        noMatchErrors.push(outputType);
+        break;
+      default:
+        throw outputType satisfies never;
+    }
+  }
+
+  if (possibleOutputTypes.length) {
+    return {
+      kind: "ok",
+      type: makeUnionAndSimplify(possibleOutputTypes),
+    };
+  }
+
+  if (noMatchErrors.length || !arityErrors.length) {
+    // at least one signature had the correct arity
+    return {
+      kind: "no-match",
+    };
+  }
+
+  // all signatures had the wrong arity
+  // TODO - if the function supports 1 or 3 args and is called with 2 args, the error will be confusing
+  // (I don't think we have any functions like that in stdlib, though)
+  return {
+    kind: "arity",
+    arity: arityErrors.flatMap((error) => error.arity), // de-dupe? (doesn't matter for now, REArityError only checks for min and max)
+  };
+}
+
+export function inferOutputTypeByLambda(lambda: Lambda, argTypes: Type[]) {
+  return inferOutputTypeByMultipleSignatures(lambda.signatures(), argTypes);
+}
+
+/**
+ * Check whether the value of type `type2` can be used in place of a variable
+ * marked with `type1`, usually as a lambda parameter.
+ *
+ * This doesn't guarantee that this substitution is safe at runtime, so this is
+ * not the same as checking for subtype/supertype relation; we intentionally
+ * don't want to implement strict type checks, we just want to make sure that
+ * there's a chance that the code won't fail.
+ */
+export function typeCanBeAssigned(type1: Type, type2: Type): boolean {
```

Comment:
This function is the heart of our type checking system. High priority for reviewing!